processor/src/test/java/com/linecorp/decaton/processor/runtime/ProcessorSubscriptionTest.java

```diff
@@ -170,4 +179,49 @@ public void testOffsetRegression() throws Exception {
         OffsetAndMetadata offset = offsets.get(tp);
         assertEquals(100L, offset.offset());
     }
+
+    @Test(timeout = 10000L)
+    public void testTerminateAsync() throws Exception {
+        TopicPartition tp = new TopicPartition("topic", 0);
+        DecatonMockConsumer consumer = new DecatonMockConsumer();
+        for (int i = 0; i < 10; i++) {
+            consumer.schedulePollTask(
+                    () -> consumer.rebalanceListener.onPartitionsAssigned(consumer.assignment()));
+        }
+        consumer.updateEndOffsets(singletonMap(tp, 10L));
+        byte[] data = {};
+        ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
+        CountDownLatch subscribed = new CountDownLatch(1);
+        CountDownLatch asyncProcessingStarted = new CountDownLatch(1);
+        DecatonProcessor<String> processor = (context, task) -> {
+            final DeferredCompletion completion = context.deferCompletion();
+            executor.schedule(completion::complete, 100, TimeUnit.MILLISECONDS);
+            asyncProcessingStarted.countDown();
+        };
+        SubscriptionScope scope = scope(tp.topic(), 9000L);
+        final ProcessorSubscription subscription = new ProcessorSubscription(
+                scope,
+                () -> consumer,
+                ProcessorsBuilder.consuming(scope.topic(),
+                                            (byte[] bytes) -> new DecatonTask<>(
+                                                    TaskMetadata.builder().build(), "dummy", bytes))
+                                 .thenProcess(processor)
+                                 .build(null),
+                scope.props(),
+                newState -> {
+                    if (newState == State.RUNNING) {
+                        subscribed.countDown();
+                    }
+                });
+        subscription.start();
+        subscribed.await();
+        consumer.rebalance(singleton(tp));
+        consumer.addRecord(new ConsumerRecord<>("topic", 0, 10, "", data));
```

Comment:
Can you add one more record with the same key so we can test that it will wait not only the tasks currently being processed but also those in queue?
To test that we have to make the above processor to block, so might be better to add another test case.

Reply:
Now that we can call initiateShutdown/awaitShutdown separately it's actually quite easy to block for the right time, I'll do it that way.